Proofread testing pyramid (#78)
sttp-book/sttp-book: chapters/pragmatic-testing/testing-pyramid.md

And a unit of work can span a single method, a whole class or multiple classes working together to achieve one single logical purpose that can be verified."

As with any testing strategy, unit testing has advantages and disadvantages.

#### Advantages

* Firstly, **unit tests are fast**.
A unit test usually takes just a couple of milliseconds to execute.
Fast tests give us the ability to test huge portions of the system in a small amount of time.
Fast, automated test suites give developers constant feedback; this fast safety net makes developers feel more comfortable/confident in performing evolutionary changes to the software system they are working on.

* Secondly, **unit tests are easy to control**. 
A unit test tests the software by giving certain parameters to a method and then comparing the return value of a method to the expected result.
The input values and expected result values are very easy to be adapted, modified, and even generated for the test.

* Finally, **unit tests are easy to write**.
Unit tests do not require complicated setup or additional work. A single unit is also
often cohesive and small, easing the job of the tester.

#### Disadvantages

> * Unit test **lack "reality"**.

Reviewer: ```suggestion
* Unit tests **lack "reality"**.
```

## System testing

Unit tests indeed do not exercise the system in its entirety (but again, that is not their goal).
To get a more realistic view of the software and thus perform more realistic tests, 
we should run the entire software system with all its 
databases, front-end apps, and any other components it is made up of.

> When we test the system in its entirity, we are doing what is called **"system testing"**.

Reviewer: ```suggestion
When we test the system in its entirety, we are doing what is called **"system testing"**.
```

In practice, instead of testing small parts of the system in isolation, system tests execute the system as a whole.
Note that an alternative name for software testing is **black box testing** because the system is some sort of black box to the testers:
we do not care or actually know what goes on inside of the system ("the black box") as long as we get the expected output for a given input.

#### Advantages

* The obvious advantage of system testing is **how realistic the tests are**.
After all,
the more realistic the tests are, the higher the chances of it actually working when shipped.

> * System tests also **capture the user's perspective** more than unit tests do.

Reviewer: ```suggestion
* System tests also **capture the user's perspective** better than unit tests do.
```

test case. This takes additional code that is needed just for automating the tests.

* Lastly, **system tests tend to become flaky**.
A flaky test is a test that presents an erratic behaviour: if you run it, it might pass or it might fail for the same configuration. Flaky tests are an important problem for software development teams. After all, having a test that might pass when there is a bug or one that might fail when there is none harms the productivity of the development team.
It is easy to imagine how a system test can become flaky. Think of a system test that exercises a web app. After clicking a button, the HTTP POST request to the web app took half a second more than usual (due to small variations we often do not control in real-life scenarios; Tomcat decided to do a full garbage collection at that very second, for example). The test was not expecting it to happen and thus, it failed. If the test is executed again, the web app might now take its usual time to respond and the test will pass on this try. There are just too many uncertainties in a system test that
can lead to unexpected behaviour.

>   We discuss _flaky tests_ more formally later in the book.

Reviewer: Please restore the "hint" tags.

<a href="http://www.youtube.com/watch?feature=player_embedded&v=5q_ZjIM_9PQ" target="_blank">
  <img src="http://img.youtube.com/vi/5q_ZjIM_9PQ/0.jpg" alt="Unit and System Tests: The Different Test Levels" width="240" height="180" border="10" />

> </a>

Reviewer: Please restore the original youtube tags.

As we saw, unit tests focus on the smallest parts of the system 
while system tests focus on the whole system at once.
However, something in between is often needed.

**Integration testing** is the test level we use when we need something more
integrated (or less isolated) than a unit test but without the need of exercising
the entire system.

Let us look at a real-world integration test example:

> Software systems commonly rely on database systems. To communicate with the database, developers often create a class whose only responsibility is to interact with this external component (think of _Data Access Objects_ - DAO - classes). These DAOs might contain complicated SQL code. Thus, a tester feels the need to
test these SQL queries.
>
> However, note that the tester does not want to test the entire system, only the integration between the DAO class and the database. The tester also does not want to test the DAO class in complete isolation; after all, the best way to know whether a SQL query works is to actually submit it to the database and see what the database returns back. This is an example of an integration test.

The goal of integration testing is to test 

> multiple components of a system together but not the whole system.
> These integration tests focus on the interaction between a few components of the system.

Reviewer: ```suggestion
multiple components of a system together but not the whole system, while focusing on the interaction between them.
```

Are they communicating correctly? What happens if component A sends message X to component B? 
Do they still present correct behaviour?

#### Advantages

The advantage of integration tests is that, while not fully isolated, devising tests 
just for a specific integration is easier than devising tests for all the components together. 
As a consequence, the effort of writing such tests

> is a bit higher than that of unit tests but lower when compared to system tests.

Reviewer: ```suggestion
is a bit higher than that of writing unit tests but lower when compared to system tests.
```

#### Disadvantages

Note that the more integrated our tests are, the harder it is to write them. 
In the example, setting up a database for the test requires effort.
Tests that involve databases usually need to:
* make use of an isolated instance of the database just for testing purposes (as you 
probably do not want your tests to mess with production data),
* update the database schema (in fast companies, database schemas are changing all the time, and the test database needs to keep up),
* put the database into a state expected by the test by adding or removing rows,
* and clean everything afterwards (so that the next tests do not fail because of the data that was left behind by the previous test).

The same effort happens to any other type of integration test you can imagine (e.g., web services, file reads and writes, etc.).

<a href="http://www.youtube.com/watch?feature=player_embedded&v=MPjQXVYqadQ" target="_blank">
  <img src="http://img.youtube.com/vi/MPjQXVYqadQ/0.jpg" alt="What is Integration Testing?" width="240" height="180" border="10" />

> </a>

Reviewer: Please restore the original youtube tags.

## The Testing Pyramid

We discussed three different test levels: unit, system, and integration. 
A question that pragmatic software developers might be asking themselves is:

_How much should I do of each?_

Testers have to decide whether to invest more in unit testing or in system testing as well as determine which components should be tested via unit testing and which components should be tested via system testing. A wrong decision might have a deep impact on the quality of the system: a wrong level might cost too much resources and might not find sufficient bugs.

While we still have no clear empirical answer to this question, practitioners have been
proposing different ways of making this decision.

One of the most famous diagrams that help us in this discussion 
is the so-called **testing pyramid**.

![Testing pyramid, extracted from Fowler's wiki](img/testing-pyramid/testing_pyramid.svg)

The diagram indicates all the test levels we discussed, plus **manual testing**. 
Note that the higher the pyramid level, the more realistic tests become and the more complex it is to devise them.

**How much should we do of each then?**

> The common practice in industry is also represented by the diagram. The size of the pyramid slice represents the amount of tests one would want of each test level. 

Reviewer: "tests" is countable, so you should use "number" rather than "amount" (just like in the original text).

> Unit test is at the bottom of the pyramid and has the largest area of them all. This means that testers should favour unit testing.

Reviewer: ```suggestion
Unit testing is at the bottom of the pyramid and has the largest area of them all. This means that testers should favour unit testing.
```

Of course, such critical parts must also be tested at other levels. Remember the _pesticide paradox_: a single technique is usually not enough to identify all the bugs.

#### **When to perform manual tests?**

Manual testing has lots of disadvantages, but is sometimes impossible to avoid. However, even in cases where automation is fully possible, manual exploratory testing can be useful. 
The Wikipedia page on [Exploratory Testing](https://en.wikipedia.org/wiki/Exploratory_testing) is well written and we point the reader to it.

On the other hand, those who apply the _testing pyramid_ try to avoid the so-called *ice-cream cone anti-pattern*. Imagine the testing pyramid but put upside down.
In this new version, manual testing has the largest area which means one should put more effort into manual testing (!!).

![Ice cream cone](img/testing-pyramid/ice_cream_cone.svg)

> At this point, we do not have to explain why fully relying on manual testing is a bad. Unfortunately, it is common to see development teams mostly rely on manual tests in their quality assurance processes. Often, these teams also have a small number of system tests. Not because they believe in their efficiency but because the system was so badly designed that unit and integration tests are simply impossible.

Reviewer: ```suggestion
At this point, we do not have to explain why fully relying on manual testing is a bad thing. Unfortunately, it is common to see development teams mostly relying on manual tests in their quality assurance processes. Often, these teams also have a small number of system tests. Not because they believe in their efficiency but because the system was so badly designed that unit and integration tests are simply impossible.
```

I changed "is a bad" to "is a bad thing", and restored "rely" back to "relying". I don't think changing it to "rely" was a grammatical improvement.

We will discuss design for testability in future chapters.

<a href="http://www.youtube.com/watch?feature=player_embedded&v=YpKxAicxasU" target="_blank">
  <img src="http://img.youtube.com/vi/YpKxAicxasU/0.jpg" alt="The Testing Pyramid" width="240" height="180" border="10" />

> </a>

Reviewer: Please restore the original youtube tags.
